src/dftbp/dftb/rshgamma.F90

```diff
@@ -0,0 +1,500 @@
+          & - (startDiag2(2) - startDiag1(2)) * (startDiag2(1) - endDiag2(1)))&
+          & / ((startDiag2(2) - endDiag2(2)) * (endDiag1(1) - startDiag1(1))&
+          & - (startDiag2(1) - endDiag2(1)) * (endDiag1(2) - startDiag1(2)))
+      ss = ((endDiag1(2) - startDiag1(2)) * (startDiag2(1) - startDiag1(1))&
+          & - (endDiag1(1) - startDiag1(1)) * (startDiag2(2) - startDiag1(2)))&
+          & / ((endDiag1(2) - startDiag1(2)) * (startDiag2(1) - endDiag2(1))&
+          & - (endDiag1(1) - startDiag1(1)) * (startDiag2(2) - endDiag2(2)))
+
+      ! check if third eq. also satisfied (3 equations but 2 variables)
+      if ((tt * (endDiag1(3) - startDiag1(3)) + ss * (startDiag2(3) - endDiag2(3)))&
+          & - (startDiag2(3) - startDiag1(3)) < 1.0e-10_dp) then
+        if ((abs(tt) >= 0.0_dp .and. abs(tt) <= 1.0_dp)&
+            & .and. (abs(ss) >= 0.0_dp .and. abs(ss) <= 1.0_dp)) then
+          intersect1 = startDiag1 + tt * (endDiag1 - startDiag1)
+          intersect2 = startDiag2 + ss * (endDiag2 - startDiag2)
+          deviation = norm2(intersect2 - intersect1)
+          if (deviation > 1.0e-10_dp) then
+            call error("Error while calculating Coulomb truncation cutoff for given cell geometry.")
+          end if
+        else
+          call error("Error while calculating Coulomb truncation cutoff for given cell geometry.")
+        end if
+      end if
+
+    end function getDiagIntersectionPoint
+
+  end function getCoulombTruncationCutoff
+
+
+  !> Workhorse for calculating the general CAM gamma integral.
+  function getCamAnalyticalGammaValue_workhorse(hubbu1, hubbu2, omega, camAlpha, camBeta, dist)&
+      & result(gamma)
+
+    !> Hubbard U's
+    real(dp), intent(in) :: hubbu1, hubbu2
+
+    !> Range-separation parameter
+    real(dp), intent(in) :: omega
+
+    !> CAM alpha and beta parameter
+    real(dp), intent(in) :: camAlpha, camBeta
+
+    !> Distance between atoms
+    real(dp), intent(in) :: dist
+
+    !> Resulting CAM gamma
+    real(dp) :: gamma
+
+    gamma =&
+        & camAlpha * getHfAnalyticalGammaValue_workhorse(hubbu1, hubbu2, dist)&
+        & + camBeta * getLrAnalyticalGammaValue_workhorse(hubbu1, hubbu2, omega, dist)
+
+  end function getCamAnalyticalGammaValue_workhorse
+
+
+  !> Workhorse for getHfAnalyticalGammaValue wrapper in hybrid module.
+  function getHfAnalyticalGammaValue_workhorse(hubbu1, hubbu2, dist) result(gamma)
+
+    !> Hubbard U's
+    real(dp), intent(in) :: hubbu1, hubbu2
+
+    !> Distance between atoms
+    real(dp), intent(in) :: dist
+
+    !> Resulting gamma
+    real(dp) :: gamma
+
+    real(dp) :: tauA, tauB
+    real(dp) :: tmp, tau
+
+    tauA = 3.2_dp * hubbu1
+    tauB = 3.2_dp * hubbu2
+
+    if (dist < tolSameDist) then
+      ! on-site case
+      if (abs(tauA - tauB) < MinHubDiff) then
+        tau = 0.5_dp * (tauA + tauB)
+        gamma = tau * 0.3125_dp
+      else
+        call error("Error(RSH-Gamma): R = 0, Ua != Ub")
```

Comment:
Pass errors upwards.

Reply:
I made the functions "error propagation ready", someone else with a better understanding of it can then pass it upwards as desired.